minion/endtoend_topic.go

```diff
@@ -0,0 +1,245 @@
+		for index := range partitions {
+			reassignedPartitions = append(reassignedPartitions, kmsg.AlterPartitionAssignmentsRequestTopicPartition{
+				Partition: int32(index),
+				Replicas:  assignedReplicas,
+			})
+		}
+
+		managamentTopicReassignment := kmsg.AlterPartitionAssignmentsRequestTopic{
+			Topic:      s.Cfg.EndToEnd.TopicManagement.Name,
+			Partitions: reassignedPartitions,
+		}
+		reassignment := kmsg.AlterPartitionAssignmentsRequest{
+			Topics: []kmsg.AlterPartitionAssignmentsRequestTopic{managamentTopicReassignment},
+		}
+		_, err := reassignment.RequestWith(ctx, s.kafkaSvc.Client)
+		if err != nil {
+			return fmt.Errorf("failed to do kmsg request on topic reassignment: %w", err)
+		}
+		return nil
+	}
+
+	return nil
+}
+
+func getTopicConfig(cfgTopic EndToEndTopicConfig) []kmsg.CreateTopicsRequestTopicConfig {
+	minISR := strconv.Itoa(cfgTopic.ReplicationFactor)
+
+	minISRConf := kmsg.CreateTopicsRequestTopicConfig{
+		Name:  "min.insync.replicas",
+		Value: &minISR,
+	}
+
+	cleanupStr := "delete"
+	cleanupPolicyConf := kmsg.CreateTopicsRequestTopicConfig{
+		Name:  "cleanup.policy",
+		Value: &cleanupStr,
+	}
+
+	retentionStr := "10000000"
+	retentionByteConf := kmsg.CreateTopicsRequestTopicConfig{
+		Name:  "retention.bytes",
+		Value: &retentionStr,
+	}
+
+	segmentStr := "1000000"
+	segmentByteConf := kmsg.CreateTopicsRequestTopicConfig{
+		Name:  "segment.bytes",
+		Value: &segmentStr,
+	}
+
+	return []kmsg.CreateTopicsRequestTopicConfig{
+		minISRConf,
+		cleanupPolicyConf,
+		retentionByteConf,
+		segmentByteConf,
+	}
+}
+
+func (s *Service) createManagementTopic(ctx context.Context, topicMetadata *kmsg.MetadataResponse) error {
+
+	s.logger.Info(fmt.Sprintf("creating topic %s for EndToEnd metrics", s.Cfg.EndToEnd.TopicManagement.Name))
+
+	cfgTopic := s.Cfg.EndToEnd.TopicManagement
+	topicConfigs := getTopicConfig(cfgTopic)
+
+	topic := kmsg.CreateTopicsRequestTopic{
+		Topic:             cfgTopic.Name,
+		NumPartitions:     int32(cfgTopic.PartitionsPerBroker),
+		ReplicationFactor: int16(cfgTopic.ReplicationFactor),
+		Configs:           topicConfigs,
+	}
+
+	// Workaround for wrong assignment on 1 ReplicationFactor with automatic assignment on topic creation, this will create the assignment manually, automatic assignment works on more than 1 RepFactor
+	// Issue: Instead of putting the number of PartitionPerBroker in One Broker/Replica, the client will assign one partition on different Broker/Replica
+	// Example for 1 RepFactor and 2 PartitionPerBroker: Instead of 2 Partitions on 1 Broker, it will put 1 Partition each on 2 Brokers
+	if cfgTopic.ReplicationFactor == 1 {
+		brokerID := topicMetadata.Brokers[0].NodeID
+		var assignment []kmsg.CreateTopicsRequestTopicReplicaAssignment
+		partitions := make([]int32, cfgTopic.PartitionsPerBroker)
+		for index := range partitions {
+			assignment = append(assignment, kmsg.CreateTopicsRequestTopicReplicaAssignment{
```

Comment:
Note it's safer to always call `Default` on each struct before using it to be future proof on new version.